docs/source/jit_language_reference_v2.rst

```diff
@@ -0,0 +1,251 @@
+cases, type annotations may be needed to overwrite auto inference, e.g., x: ``Optional[int] = None``. Note that it is always safe to type annotate a local variable
+or data attribute even if its type can be automatically inferred. But the annotated type must be congruent with TorchScript’s type checking.
+
+When a parameter, local variable, or data attribute is not type annotated and its type cannot be automatically inferred, TorchScript assumes it to be a
+default type of ``TensorType``.
+
+Annotate function signature
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Since parameter may not be automatically inferred from the body of the function (including both functions and methods), they need to be type annotated,
+otherwise they assume the default type ``TensorType``.
+
+TorchScript supports two styles for method and function signature type annotation:
+
+* **Python3-style** annotates types directly on the signature. As such, it allows individual parameters be left unannotated
+(whose type will be the default type of ``TensorType``) , or the return type be left unannotated (whose type will be automatically inferred).
+
+::
+
+    Python3Annotation := "def" Identifier [ "(" ParamAnnot* ")" ] [ReturnAnnot] ":"
+                         FuncOrMethodBody
+    ParamAnnot := Identifier [ ":" TSType ] ","
+    ReturnAnnot := "->" TSType
+
+Note that using Python3 style, the type of ``self`` is automatically inferred and should not be annotated.
+
+* **Mypy style** annotates types as a comment right below the function/method declaration. In the My-Py style, since parameter names do not appear
+in the annotation, all parameters have to be annotated.
+
+::
+
+    MyPyAnnotation := "# type:" "(" ParamAnnot* ")" [ ReturnAnnot ]
+    ParamAnnot := TSType ","
+    ReturnAnnot := "->" TSType
+
+**Example 1**
+
+In this example, ``a`` is not annotated and assumes the default type of ``TensorType``, ``b`` is annotated as type ``int``, and the return type is not
+annotated and is automatically inferred as type ``TensorType``.
+
+::
+
+    import torch
+
+    def f(a, b: int):
+        return a+b
+
+    m = torch.jit.script(f)
+    print("TorchScript:", m(torch.ones([6]), 100))
+
+**Example 2**
+
+The following code snippet gives an example of using mypy style annotation. Note that parameters or return values must be annotated even if some of
+them assume the default type.
+
+::
+
+    import torch
+
+    def f(a, b):
+        # type: (torch.Tensor, int) → torch.Tensor
+        return a+b
+
+    m = torch.jit.script(f)
+    print("TorchScript:", m(torch.ones([6]), 100))
+
+**Example 3**
+
+This example caused a compilation error because it type annotates ``self``. The solution is to remove the type annotation and let its type be automatically
+inferred.
+
+::
+
+    import torch
+
+    @torch.jit.script
+    class MyClass(object):
+    # ERROR: do not annotate the type of self
+    def __init__(self: MyClass, x: int):
```

Comment:
This isn't the best example because the failure is caused by "using MyClass before it is fully defined in Python yet".